File: Traditional-Important-Algorithms/exercises_py/03_concurrency/con09_coroutine_scheduling.py
# I AM NOT DONE

"""
con09_coroutine_scheduling.py

Coroutines are functions that can suspend execution and resume later,
allowing for cooperative multitasking. A coroutine scheduler manages
multiple coroutines, switching between them when they yield.

Your task: Implement a coroutine scheduler that can suspend and resume
coroutine execution.

Key concepts:
- Yield points: Where coroutines can suspend
- Resumption: Continuing from where a coroutine left off
- Cooperative scheduling: Coroutines control when they yield
- State preservation: Maintaining coroutine state across yields
"""

from collections import deque
from enum import Enum
from typing import List, Optional
from dataclasses import dataclass
import unittest


CoroutineId = int


class YieldType(Enum):
    """Types of yields."""
    YIELD = "yield"
    YIELD_VALUE = "yield_value"
    AWAIT = "await"


class CoroutineState(Enum):
    """Coroutine execution states."""
    READY = "ready"
    RUNNING = "running"
    SUSPENDED = "suspended"
    COMPLETED = "completed"


@dataclass
class Instruction:
    """Instruction for coroutine execution."""
    execute_for: int  # Time units to execute
    action: str  # 'yield' or 'return'
    yield_type: Optional[YieldType] = None
    yield_value: Optional[int] = None
    await_coroutine: Optional[CoroutineId] = None
    return_value: Optional[int] = None


class Coroutine:
    """Represents a coroutine with instructions."""

    def __init__(self, coroutine_id: CoroutineId, instructions: List[Instruction]):
        self.id = coroutine_id
        self.state = CoroutineState.READY
        self.instructions = deque(instructions)
        self.execution_time = 0

    @staticmethod
    def simple_yielding(coroutine_id: CoroutineId, total_time: int, yield_interval: int):
        """
        TODO: Create a coroutine that yields every yield_interval units.

        Args:
            coroutine_id: Coroutine ID
            total_time: Total execution time
            yield_interval: Interval between yields

        Returns:
            A Coroutine that yields periodically
        """
        pass  # TODO: Implement this

    def is_complete(self) -> bool:
        """Check if coroutine has completed."""
        return self.state == CoroutineState.COMPLETED

    def can_run(self) -> bool:
        """Check if coroutine can run."""
        return self.state == CoroutineState.READY


class EventType(Enum):
    """Scheduler event types."""
    STARTED = "started"
    RESUMED = "resumed"
    YIELDED = "yielded"
    COMPLETED = "completed"


@dataclass
class SchedulerEvent:
    """Records scheduler events."""
    time: int
    coroutine_id: CoroutineId
    event: EventType
    yield_type: Optional[YieldType] = None
    return_value: Optional[int] = None


class CoroutineScheduler:
    """Scheduler for managing coroutines."""

    def __init__(self):
        self.coroutines: List[Coroutine] = []
        self.ready_queue = deque()
        self.waiting: List[Tuple[CoroutineId, CoroutineId]] = []  # (waiting_id, waiting_for_id)
        self.current_time = 0
        self.events: List[SchedulerEvent] = []

    def spawn(self, coroutine: Coroutine) -> CoroutineId:
        """
        TODO: Add a coroutine to the scheduler.

        Steps:
        1. Add to coroutines list
        2. Add to ready queue
        3. Return its ID

        Args:
            coroutine: The coroutine to add

        Returns:
            The coroutine ID
        """
        pass  # TODO: Implement this

    def _log_event(self, coroutine_id: CoroutineId, event: EventType,
                   yield_type: Optional[YieldType] = None,
                   return_value: Optional[int] = None) -> None:
        """Log a scheduler event."""
        self.events.append(SchedulerEvent(self.current_time, coroutine_id, event,
                                          yield_type, return_value))

    def _resume_waiting_coroutines(self, completed_id: CoroutineId) -> None:
        """
        TODO: Resume coroutines that were waiting for the completed coroutine.

        Steps:
        1. Find all coroutines waiting for completed_id
        2. Move them back to ready queue
        3. Update their state to Ready
        4. Remove them from waiting list

        Args:
            completed_id: ID of the completed coroutine
        """
        pass  # TODO: Implement this

    def run_next(self) -> Optional[CoroutineId]:
        """
        TODO: Run the next ready coroutine until it yields or completes.

        Steps:
        1. Get next coroutine from ready queue
        2. Update its state to Running
        3. Execute its next instruction
        4. Handle the action (Yield or Return)
        5. Update current_time
        6. Log appropriate events
        7. Return the coroutine ID that ran

        Returns:
            The coroutine ID that ran, or None
        """
        pass  # TODO: Implement this

    def run_all(self) -> None:
        """Run all coroutines until completion."""
        while len(self.ready_queue) > 0 or len(self.waiting) > 0:
            if len(self.ready_queue) == 0:
                # Deadlock detection
                if len(self.waiting) > 0:
                    break
            self.run_next()

    def get_coroutine(self, coroutine_id: CoroutineId) -> Optional[Coroutine]:
        """Get a coroutine by ID."""
        for coro in self.coroutines:
            if coro.id == coroutine_id:
                return coro
        return None

    def current_time_value(self) -> int:
        """Get the current time."""
        return self.current_time

    def get_events(self) -> List[SchedulerEvent]:
        """Get all logged events."""
        return self.events

    def all_complete(self) -> bool:
        """Check if all coroutines have completed."""
        return all(c.is_complete() for c in self.coroutines)


class TestCoroutineScheduler(unittest.TestCase):
    """Test cases for Coroutine Scheduler."""

    def test_single_coroutine_no_yield(self):
        """Test single coroutine without yields."""
        scheduler = CoroutineScheduler()

        coroutine = Coroutine(0, [
            Instruction(10, 'return', return_value=42),
        ])

        scheduler.spawn(coroutine)
        scheduler.run_all()

        coro = scheduler.get_coroutine(0)
        self.assertEqual(coro.state, CoroutineState.COMPLETED)
        self.assertEqual(scheduler.current_time_value(), 10)

    def test_coroutine_with_yields(self):
        """Test coroutine with yields."""
        scheduler = CoroutineScheduler()

        coroutine = Coroutine(0, [
            Instruction(5, 'yield', yield_type=YieldType.YIELD),
            Instruction(5, 'return', return_value=100),
        ])

        scheduler.spawn(coroutine)
        scheduler.run_all()

        events = scheduler.get_events()
        self.assertTrue(any(e.event == EventType.YIELDED and
                           e.yield_type == YieldType.YIELD for e in events))
        self.assertTrue(any(e.event == EventType.COMPLETED and
                           e.return_value == 100 for e in events))

    def test_multiple_coroutines(self):
        """Test multiple coroutines executing."""
        scheduler = CoroutineScheduler()

        scheduler.spawn(Coroutine.simple_yielding(0, 10, 5))
        scheduler.spawn(Coroutine.simple_yielding(1, 10, 5))

        scheduler.run_all()

        self.assertTrue(scheduler.all_complete())

        # Both should have run
        events = scheduler.get_events()
        self.assertTrue(any(e.coroutine_id == 0 for e in events))
        self.assertTrue(any(e.coroutine_id == 1 for e in events))

    def test_round_robin_execution(self):
        """Test round-robin execution of coroutines."""
        scheduler = CoroutineScheduler()

        scheduler.spawn(Coroutine.simple_yielding(0, 15, 5))
        scheduler.spawn(Coroutine.simple_yielding(1, 15, 5))

        scheduler.run_all()

        events = scheduler.get_events()

        # Events should alternate between coroutines
        last_id = None
        switches = 0

        for event in events:
            if event.event in (EventType.STARTED, EventType.RESUMED):
                if last_id is not None and last_id != event.coroutine_id:
                    switches += 1
                last_id = event.coroutine_id

        self.assertGreater(switches, 0, "Coroutines should interleave execution")

    def test_await_coroutine(self):
        """Test awaiting another coroutine."""
        scheduler = CoroutineScheduler()

        coroutine1 = Coroutine(0, [
            Instruction(10, 'return', return_value=42),
        ])

        coroutine2 = Coroutine(1, [
            Instruction(5, 'yield', yield_type=YieldType.AWAIT, await_coroutine=0),
            Instruction(5, 'return', return_value=100),
        ])

        scheduler.spawn(coroutine1)
        scheduler.spawn(coroutine2)
        scheduler.run_all()

        # Both should complete
        self.assertTrue(scheduler.all_complete())

        events = scheduler.get_events()
        # Coroutine 1 should complete before coroutine 2 resumes
        coro1_complete_events = [e for e in events
                                 if e.coroutine_id == 0 and e.event == EventType.COMPLETED]
        coro2_resume_events = [e for e in events
                              if e.coroutine_id == 1 and e.event == EventType.RESUMED]

        if coro1_complete_events and coro2_resume_events:
            complete_time = coro1_complete_events[0].time
            resume_time = coro2_resume_events[0].time
            self.assertLessEqual(complete_time, resume_time)

    def test_immediate_completion(self):
        """Test coroutine that completes immediately."""
        scheduler = CoroutineScheduler()

        coroutine = Coroutine(0, [
            Instruction(0, 'return', return_value=1),
        ])

        scheduler.spawn(coroutine)
        scheduler.run_all()

        self.assertTrue(scheduler.all_complete())
        self.assertEqual(scheduler.current_time_value(), 0)

    def test_multiple_yields(self):
        """Test coroutine with multiple yields."""
        scheduler = CoroutineScheduler()

        coroutine = Coroutine(0, [
            Instruction(1, 'yield', yield_type=YieldType.YIELD),
            Instruction(1, 'yield', yield_type=YieldType.YIELD),
            Instruction(1, 'yield', yield_type=YieldType.YIELD),
            Instruction(1, 'return', return_value=0),
        ])

        scheduler.spawn(coroutine)
        scheduler.run_all()

        events = scheduler.get_events()
        yield_count = sum(1 for e in events if e.event == EventType.YIELDED)

        self.assertEqual(yield_count, 3)

    def test_empty_scheduler(self):
        """Test running empty scheduler."""
        scheduler = CoroutineScheduler()

        result = scheduler.run_next()
        self.assertIsNone(result)


if __name__ == '__main__':
    unittest.main()

File: Traditional-Important-Algorithms/exercises/03_concurrency/con09_coroutine_scheduling.rs
// con09_coroutine_scheduling.rs
//
// Coroutines are functions that can suspend execution and resume later,
// allowing for cooperative multitasking. A coroutine scheduler manages
// multiple coroutines, switching between them when they yield.
//
// Your task: Implement a coroutine scheduler that can suspend and resume
// coroutine execution.
//
// Key concepts:
// - Yield points: Where coroutines can suspend
// - Resumption: Continuing from where a coroutine left off
// - Cooperative scheduling: Coroutines control when they yield
// - State preservation: Maintaining coroutine state across yields

// I AM NOT DONE

use std::collections::VecDeque;

pub type CoroutineId = usize;

#[derive(Debug, Clone, PartialEq)]
pub enum YieldType {
    Yield,          // Simple yield to scheduler
    YieldValue(i32), // Yield with a value
    Await(CoroutineId), // Wait for another coroutine to complete
}

#[derive(Debug, Clone, PartialEq)]
pub enum CoroutineState {
    Ready,
    Running,
    Suspended(YieldType),
    Completed(Option<i32>),
}

#[derive(Debug, Clone)]
pub struct Instruction {
    pub execute_for: u32,  // Time units to execute
    pub then: Action,       // What to do after execution
}

#[derive(Debug, Clone)]
pub enum Action {
    Yield(YieldType),
    Return(i32),
}

pub struct Coroutine {
    pub id: CoroutineId,
    pub state: CoroutineState,
    instructions: VecDeque<Instruction>,
    execution_time: u32,
}

impl Coroutine {
    pub fn new(id: CoroutineId, instructions: Vec<Instruction>) -> Self {
        Self {
            id,
            state: CoroutineState::Ready,
            instructions: instructions.into_iter().collect(),
            execution_time: 0,
        }
    }

    pub fn simple_yielding(id: CoroutineId, total_time: u32, yield_interval: u32) -> Self {
        // TODO: Create a coroutine that yields every yield_interval units
        // and completes after total_time units
        todo!()
    }

    pub fn is_complete(&self) -> bool {
        matches!(self.state, CoroutineState::Completed(_))
    }

    pub fn can_run(&self) -> bool {
        matches!(self.state, CoroutineState::Ready)
    }
}

#[derive(Debug, Clone)]
pub struct SchedulerEvent {
    pub time: u32,
    pub coroutine_id: CoroutineId,
    pub event: EventType,
}

#[derive(Debug, Clone, PartialEq)]
pub enum EventType {
    Started,
    Resumed,
    Yielded(YieldType),
    Completed(Option<i32>),
}

pub struct CoroutineScheduler {
    coroutines: Vec<Coroutine>,
    ready_queue: VecDeque<CoroutineId>,
    waiting: Vec<(CoroutineId, CoroutineId)>, // (waiting_id, waiting_for_id)
    current_time: u32,
    events: Vec<SchedulerEvent>,
}

impl CoroutineScheduler {
    pub fn new() -> Self {
        Self {
            coroutines: Vec::new(),
            ready_queue: VecDeque::new(),
            waiting: Vec::new(),
            current_time: 0,
            events: Vec::new(),
        }
    }

    pub fn spawn(&mut self, coroutine: Coroutine) -> CoroutineId {
        // TODO: Add a coroutine to the scheduler
        // 1. Add to coroutines vector
        // 2. Add to ready queue
        // 3. Return its ID
        todo!()
    }

    fn log_event(&mut self, coroutine_id: CoroutineId, event: EventType) {
        self.events.push(SchedulerEvent {
            time: self.current_time,
            coroutine_id,
            event,
        });
    }

    fn resume_waiting_coroutines(&mut self, completed_id: CoroutineId) {
        // TODO: Resume coroutines that were waiting for the completed coroutine
        // 1. Find all coroutines waiting for completed_id
        // 2. Move them back to ready queue
        // 3. Update their state to Ready
        // 4. Remove them from waiting list
        todo!()
    }

    pub fn run_next(&mut self) -> Option<CoroutineId> {
        // TODO: Run the next ready coroutine until it yields or completes
        // 1. Get next coroutine from ready queue
        // 2. Update its state to Running
        // 3. Execute its next instruction
        // 4. Handle the action (Yield or Return)
        // 5. Update current_time
        // 6. Log appropriate events
        // 7. Return the coroutine ID that ran
        todo!()
    }

    pub fn run_all(&mut self) {
        // TODO: Run all coroutines until completion
        while !self.ready_queue.is_empty() || !self.waiting.is_empty() {
            if self.ready_queue.is_empty() {
                // Deadlock detection: if waiting list is not empty but ready queue is,
                // we have a deadlock
                if !self.waiting.is_empty() {
                    break;
                }
            }
            self.run_next();
        }
    }

    pub fn get_coroutine(&self, id: CoroutineId) -> Option<&Coroutine> {
        self.coroutines.iter().find(|c| c.id == id)
    }

    pub fn current_time(&self) -> u32 {
        self.current_time
    }

    pub fn get_events(&self) -> &[SchedulerEvent] {
        &self.events
    }

    pub fn all_complete(&self) -> bool {
        self.coroutines.iter().all(|c| c.is_complete())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_single_coroutine_no_yield() {
        let mut scheduler = CoroutineScheduler::new();

        let coroutine = Coroutine::new(
            0,
            vec![Instruction {
                execute_for: 10,
                then: Action::Return(42),
            }],
        );

        scheduler.spawn(coroutine);
        scheduler.run_all();

        let coro = scheduler.get_coroutine(0).unwrap();
        assert_eq!(coro.state, CoroutineState::Completed(Some(42)));
        assert_eq!(scheduler.current_time(), 10);
    }

    #[test]
    fn test_coroutine_with_yields() {
        let mut scheduler = CoroutineScheduler::new();

        let coroutine = Coroutine::new(
            0,
            vec![
                Instruction {
                    execute_for: 5,
                    then: Action::Yield(YieldType::Yield),
                },
                Instruction {
                    execute_for: 5,
                    then: Action::Return(100),
                },
            ],
        );

        scheduler.spawn(coroutine);
        scheduler.run_all();

        let events = scheduler.get_events();
        assert!(events.iter().any(|e| e.event == EventType::Yielded(YieldType::Yield)));
        assert!(events.iter().any(|e| e.event == EventType::Completed(Some(100))));
    }

    #[test]
    fn test_multiple_coroutines() {
        let mut scheduler = CoroutineScheduler::new();

        scheduler.spawn(Coroutine::simple_yielding(0, 10, 5));
        scheduler.spawn(Coroutine::simple_yielding(1, 10, 5));

        scheduler.run_all();

        assert!(scheduler.all_complete());

        // Both should have run
        let events = scheduler.get_events();
        assert!(events.iter().any(|e| e.coroutine_id == 0));
        assert!(events.iter().any(|e| e.coroutine_id == 1));
    }

    #[test]
    fn test_round_robin_execution() {
        let mut scheduler = CoroutineScheduler::new();

        scheduler.spawn(Coroutine::simple_yielding(0, 15, 5));
        scheduler.spawn(Coroutine::simple_yielding(1, 15, 5));

        scheduler.run_all();

        let events = scheduler.get_events();

        // Events should alternate between coroutines
        let mut last_id = None;
        let mut switches = 0;

        for event in events {
            if matches!(event.event, EventType::Started | EventType::Resumed) {
                if let Some(last) = last_id {
                    if last != event.coroutine_id {
                        switches += 1;
                    }
                }
                last_id = Some(event.coroutine_id);
            }
        }

        assert!(switches > 0, "Coroutines should interleave execution");
    }

    #[test]
    fn test_yield_value() {
        let mut scheduler = CoroutineScheduler::new();

        let coroutine = Coroutine::new(
            0,
            vec![
                Instruction {
                    execute_for: 3,
                    then: Action::Yield(YieldType::YieldValue(42)),
                },
                Instruction {
                    execute_for: 3,
                    then: Action::Return(100),
                },
            ],
        );

        scheduler.spawn(coroutine);
        scheduler.run_all();

        let events = scheduler.get_events();
        assert!(events.iter().any(|e| e.event == EventType::Yielded(YieldType::YieldValue(42))));
    }

    #[test]
    fn test_await_coroutine() {
        let mut scheduler = CoroutineScheduler::new();

        let coroutine1 = Coroutine::new(
            0,
            vec![Instruction {
                execute_for: 10,
                then: Action::Return(42),
            }],
        );

        let coroutine2 = Coroutine::new(
            1,
            vec![
                Instruction {
                    execute_for: 5,
                    then: Action::Yield(YieldType::Await(0)),
                },
                Instruction {
                    execute_for: 5,
                    then: Action::Return(100),
                },
            ],
        );

        scheduler.spawn(coroutine1);
        scheduler.spawn(coroutine2);
        scheduler.run_all();

        // Both should complete
        assert!(scheduler.all_complete());

        let events = scheduler.get_events();
        // Coroutine 1 should complete before coroutine 2 resumes
        let coro1_complete_time = events
            .iter()
            .find(|e| e.coroutine_id == 0 && matches!(e.event, EventType::Completed(_)))
            .map(|e| e.time);

        let coro2_resume_time = events
            .iter()
            .filter(|e| e.coroutine_id == 1 && e.event == EventType::Resumed)
            .map(|e| e.time)
            .next();

        if let (Some(complete), Some(resume)) = (coro1_complete_time, coro2_resume_time) {
            assert!(complete <= resume, "Coroutine 2 should resume after coroutine 1 completes");
        }
    }

    #[test]
    fn test_immediate_completion() {
        let mut scheduler = CoroutineScheduler::new();

        let coroutine = Coroutine::new(
            0,
            vec![Instruction {
                execute_for: 0,
                then: Action::Return(1),
            }],
        );

        scheduler.spawn(coroutine);
        scheduler.run_all();

        assert!(scheduler.all_complete());
        assert_eq!(scheduler.current_time(), 0);
    }

    #[test]
    fn test_multiple_yields() {
        let mut scheduler = CoroutineScheduler::new();

        let coroutine = Coroutine::new(
            0,
            vec![
                Instruction {
                    execute_for: 1,
                    then: Action::Yield(YieldType::Yield),
                },
                Instruction {
                    execute_for: 1,
                    then: Action::Yield(YieldType::Yield),
                },
                Instruction {
                    execute_for: 1,
                    then: Action::Yield(YieldType::Yield),
                },
                Instruction {
                    execute_for: 1,
                    then: Action::Return(0),
                },
            ],
        );

        scheduler.spawn(coroutine);
        scheduler.run_all();

        let events = scheduler.get_events();
        let yield_count = events
            .iter()
            .filter(|e| matches!(e.event, EventType::Yielded(_)))
            .count();

        assert_eq!(yield_count, 3);
    }
}
